# Introduction into Data Science - Assignment Part II

This is the second part of the assignment in IDS 2023/2024.

This part of the assignment consists of five questions — each of these questions is contained in a separate Jupyter notebook:
- [Question 1: Data Preprocessing](Q1_Preprocessing_Visualization.ipynb)
- [Question 2: Association Rules](Q2_Frequent_Itemsets_Association_Rules.ipynb)
- [Question 3: Process Mining](Q3_Process_Mining.ipynb)
- [Question 4: Text Mining](Q4_Text_Mining.ipynb)
- [Question 5: Big Data](Q5_Big_Data.ipynb)

Additional required files are in two folders.
- [datasets](datasets/)
- [scripts](scripts/)

Please use the provided notebook to work on the questions. When you are done, upload your version of each of the notebooks to Moodle. Your submission will, therefore, consist of five jupyter notebook and _no_ additional file. Any additionally provided files will not be considered in grading.
Enter your commented Python code and answers in the corresponding cells. Make sure to answer all questions in a clear and explicit manner and discuss your outputs. _Please do not change the general structure of this notebook_. You can, however, add additional markdown or code cells if necessary. Please **DO NOT CLEAR THE OUTPUT** of the notebook you are submitting! Additionally, please ensure that the code in the notebook runs if placed in the same folder as all of the provided files, delivering the same outputs as the ones you submit in the notebook. This includes being runnable in the bundled conda environment.

*Please make sure to include the names and matriculation numbers of all group members in the provided slots in each of the notebooks.* If a name or a student id is missing, the student will not receive any points.

Hint 1: **Plan your time wisely.** A few parts of this assignment may take some time to run. It might be necessary to consider time management when you plan your group work. Also, do not attempt to upload your assignment at the last minute before the deadline. This often does not work, and you will miss the deadline. Late submissions will not be considered.

Hint 2: RWTHMoodle allows multiple submissions, with every new submission overwriting the previous one. **Partial submissions are possible and encouraged.** This might be helpful in case of technical issues with RWTHMoodle, which may occur close to the deadline.

Hint 3: As a technical note. Some IDEs such as DataSpell may automatically strip jupyter notebook cell metadata. If you are able, please re-add it from the source notebooks before submission. This is necessary for our grading.

Enter your group number and members with matriculation numbers below.

In [1]:
GROUP_NO = 47 # group number
GROUP_MEMBERS = {
    460101: "Feiyu Wang", # mat. no. : name,
    460790: "Malte Schörken",
    461277: "Chenmingxuan Feng",
}

---

In [2]:
# required imports
# please do not edit!
import csv
from mlxtend.preprocessing import TransactionEncoder
import pandas as pd
from mlxtend.frequent_patterns import apriori
import datetime
from mlxtend.frequent_patterns import association_rules as arule
import pandas as pd
from mlxtend.frequent_patterns import fpgrowth

# Question 2: Frequent Item Sets and Association Rules (13 points)

In this question, you work with transaction data of the customer's visits to the store.

### a) 
Load the transactions from the csv-file called **q2_store_transactions.csv** into a variable called `groceries`. The variable should be a list and each row in the csv-file should be represented as a list within this list.

In [3]:
import csv

# YOUR CODE HERE
groceries = []

with open('datasets/q2_store_transactions.csv', 'r') as file:
    reader = csv.reader(file)
    next(reader, None) 
    for row in reader:
        groceries.append(row)

In [4]:
# Please leave this cell empty - used for grading.

### b) 
Transform the entries from the list to a binary matrix using an object of *TransactionEncoder* as introduced in the exercise. Name the resulting dataframe `itemset_matrix` and display the first 20 rows.

In [5]:
from mlxtend.preprocessing import TransactionEncoder
import pandas as pd

# YOUR CODE HERE
te = TransactionEncoder().fit(groceries)
matrix1 = te.transform(groceries)
itemset_matrix = pd.DataFrame(matrix1, columns=te.columns_)
itemset_matrix.head(20)

,asparagus,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,black tea,blueberries,...,turkey,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,True,False,False,False,False,False,...,True,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
5,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,True,False,False,False,False
6,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
8,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
9,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


In [6]:
# Please leave this cell empty - used for grading.

In [7]:
# Please leave this cell empty - used for grading.

### c) 
Find all frequent itemsets with a **support of at least 0.03** using the Apriori algorithm and save them in a variable called `frequent_itemsets`. Display the resulting itemsets and the processing time (in milliseconds) required to detect them. 

In [8]:
from mlxtend.frequent_patterns import apriori
import datetime

# YOUR CODE HERE
start_time = datetime.datetime.now()

frequent_itemsets = apriori(itemset_matrix, min_support=0.03, use_colnames=True)

end_time = datetime.datetime.now()

processing_time_ms = (end_time - start_time).total_seconds() * 1000
print("Frequent Itemsets:")
print(frequent_itemsets)
print("\nProcessing Time (ms):", processing_time_ms)

Frequent Itemsets:
     support                            itemsets
0   0.033200                           (avocado)
1   0.033733                          (brownies)
2   0.087200                           (burgers)
3   0.030133                            (butter)
4   0.081067                              (cake)
5   0.046800                         (champagne)
6   0.060000                           (chicken)
7   0.163867                         (chocolate)
8   0.080400                           (cookies)
9   0.051067                       (cooking oil)
10  0.031733                    (cottage cheese)
11  0.179733                              (eggs)
12  0.079333                          (escalope)
13  0.170933                      (french fries)
14  0.043067                       (fresh bread)
15  0.063200                   (frozen smoothie)
16  0.095333                 (frozen vegetables)
17  0.052400                     (grated cheese)
18  0.132000                         (green tea)
1

In [9]:
# Please leave this cell empty - used for grading.

### d)
Find the most frequent itemsets containing **more than one product** and a **support of more than 0.04** using the Apriori algorithm. Store them in a variable called `frequent_itemsets_filtered` and show the sets in your output.

In [10]:
# YOUR CODE HERE
frequent_itemsets_filtered = frequent_itemsets[(frequent_itemsets['itemsets'].apply(len) > 1) & (frequent_itemsets['support'] > 0.04)]
frequent_itemsets_filtered

,support,itemsets
39,0.052667,"(chocolate, mineral water)"
43,0.050933,"(mineral water, eggs)"
48,0.040933,"(mineral water, ground beef)"
50,0.048000,"(milk, mineral water)"
53,0.059733,"(mineral water, spaghetti)"


In [11]:
# Please leave this cell empty - used for grading.

In [12]:
# Please leave this cell empty - used for grading.

In [13]:
# Please leave this cell empty - used for grading.

### e)
Find all association rules in the data that have a **confidence of at least 0.3** and a **minimum lift of 1.2**. Create and show a dataframe `association_rules` listing the antecedents, consequents, support, confidence, and lift of each of these discovered rules. How do you interpret the quality of the discovered rules?

In [14]:
from mlxtend.frequent_patterns import association_rules as arule
import pandas as pd

# YOUR CODE HERE
assoc_rules = arule(frequent_itemsets, metric='confidence', min_threshold=0.3)
association_rules = assoc_rules.loc[(assoc_rules['lift'] >= 1.2)]
print(association_rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])

itemsets_with_water = 0
for itemset in groceries:
    if 'mineral water' in itemset:
        itemsets_with_water += 1
print(itemsets_with_water/len(groceries))

           antecedents      consequents   support  confidence      lift
0          (chocolate)  (mineral water)  0.052667    0.321400  1.348907
1  (frozen vegetables)  (mineral water)  0.035733    0.374825  1.573133
2        (ground beef)  (mineral water)  0.040933    0.416554  1.748266
3        (ground beef)      (spaghetti)  0.039200    0.398915  2.290857
4               (milk)  (mineral water)  0.048000    0.370370  1.554436
5           (pancakes)  (mineral water)  0.033733    0.354839  1.489250
6          (spaghetti)  (mineral water)  0.059733    0.343032  1.439698
0.23826666666666665


In [15]:
# Please leave this cell empty - used for grading.

__Student Answer:__ Almost all of the rules with confidence >= 0.3 and lift >=1.2 include {mineral water} in the consequents. {mineral water} is contained in about 23.84% of the shopping baskets. This explains why {mineral water} appears so frequently in association rules.

The forth rule: {ground beef} implies {spaghetti}, is the only association rule not including mineral water. It has the highest lift of about 2.29. This is the only rule that can be considered a high-quality rule with a strong association between the items.

### f) 
Find all frequent itemsets with a **support of at least 0.03** using **FP-Growth** and save them in a variable called `fp_frequent_itemsets`. Display the resulting itemsets and the processing time (in milliseconds) required to detect them. 

In [16]:
from mlxtend.frequent_patterns import fpgrowth

# YOUR CODE HERE
start_time = datetime.datetime.now()

fp_frequent_itemsets = fpgrowth(itemset_matrix, min_support=0.03, use_colnames=True)

end_time = datetime.datetime.now()
processing_time_ms = (end_time - start_time).total_seconds() * 1000

print("Frequent Itemsets using FP-Growth:")
print(fp_frequent_itemsets)
print("\nProcessing Time (ms):", processing_time_ms)

Frequent Itemsets using FP-Growth:
     support                            itemsets
0   0.179733                              (eggs)
1   0.087200                           (burgers)
2   0.062533                            (turkey)
3   0.033200                           (avocado)
4   0.238267                     (mineral water)
5   0.132000                         (green tea)
6   0.129600                              (milk)
7   0.058533                  (whole wheat rice)
8   0.076400                    (low fat yogurt)
9   0.170933                      (french fries)
10  0.050533                              (soup)
11  0.174133                         (spaghetti)
12  0.095333                 (frozen vegetables)
13  0.080400                           (cookies)
14  0.051067                       (cooking oil)
15  0.046800                         (champagne)
16  0.042400                            (salmon)
17  0.163867                         (chocolate)
18  0.071333                      

In [17]:
# Please leave this cell empty - used for grading.

### g)
Using the itemsets identified by **FP-Growth**: Find all association rules in the data that have a **confidence of at least 0.3** and a **minimum lift of 1.2**. Create and show a dataframe `fp_association_rules` listing the antecedents, consequents, support, confidence, and lift of each of these discovered rules.

In [18]:
# YOUR CODE HERE
fp_association_rules = arule(fp_frequent_itemsets, metric='confidence', min_threshold=0.3)
fp_association_rules = fp_association_rules.loc[fp_association_rules['lift'] >= 1.2]

fp_association_rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']]

,antecedents,consequents,support,confidence,lift
0,(milk),(mineral water),0.048000,0.370370,1.554436
1,(spaghetti),(mineral water),0.059733,0.343032,1.439698
2,(frozen vegetables),(mineral water),0.035733,0.374825,1.573133
3,(chocolate),(mineral water),0.052667,0.321400,1.348907
4,(pancakes),(mineral water),0.033733,0.354839,1.489250
5,(ground beef),(mineral water),0.040933,0.416554,1.748266
6,(ground beef),(spaghetti),0.039200,0.398915,2.290857


In [19]:
# Please leave this cell empty - used for grading.

### h) 
You would like to compare the apriori algorithms and FP-Growth.

i) Both algorithms use the same data (transaction data) as an input and provide association rules as an output. How do the algorithms differ in the way they identify association rules?


__Student Answer:__ 

Apriori:
- Apriori exploits the fact that adding elements can only make an itemset less frequent and removing items can only make an itemset more frequent. 
- Apriori uses candidate generation where frequent subsets are extended one item at a time.
- Each candidate needs to be tested against the whole dataset. The support has to be calculated. 
- Breadth-first

FP-Growth:
- Builds an FP-Tree based on the frequency of each item
- Basically works in the 4 steps:
1. Count overall frequency and remove infrequent items.
2. Sort all projected instances in order of global frequency
3. Build a tree adding the projected instances one-by-one
4. Prune the tree
- Avoids generating many candidates
- Depth-first
- Big Advantage: Only two passes through the data are needed, faster algorithm

ii) Consider your results of the previous tasks. Do the two algorithms provide the same association rules? Is this always the case?

__Student Answer:__ The two algorithms do provide the same association rules. When minimum confidence and lift are the same the two algorithms should always output the same association rules. They use different methods to calculate frequent itemsets but these itemsets are deterministic. So are the association rules. The only difference is the rounding in the values because of computer inaccuracy. So there might be some very minor differences.

iii) Compare the processing time for finding the frequent itemsets tasks using the apriori algorithm and FG-Growth. What do you notice? Is this the result you expected? Briefly explain your answers.


__Student Answer:__ In this case, apriori algorithm takes less time than FP-Growth.
The observed result is rather unexpected. On larger datasets FP-Growth should take less time than Apriori because of the advantages discussed above. It seems like the dataset was too small for FP-Growth to be faster. Both algorithms are very fast on this dataset.